extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/WorkerSketchFetcher.java

```diff
@@ -0,0 +1,253 @@
+        List<String> workerTaskIds,
+        Iterator<Map.Entry<Long, Set<Integer>>> timeSegmentVsWorkerIdIterator
+    )
+    {
+      this.finalPartitionBoundries = new ArrayList<>();
+      this.stageDefinition = stageDefinition;
+      this.workerTaskIds = workerTaskIds;
+      this.timeSegmentVsWorkerIdIterator = timeSegmentVsWorkerIdIterator;
+      this.partitionFuture = new CompletableFuture<>();
+    }
+
+    public void submitFetchingTasksForNextTimeChunk()
+    {
+      if (!timeSegmentVsWorkerIdIterator.hasNext()) {
+        partitionFuture.complete(Either.value(new ClusterByPartitions(finalPartitionBoundries)));
+      } else {
+        Map.Entry<Long, Set<Integer>> entry = timeSegmentVsWorkerIdIterator.next();
+        Long timeChunk = entry.getKey();
+        Set<Integer> workerIdsWithTimeChunk = entry.getValue();
+        ClusterByStatisticsCollector mergedStatisticsCollector = stageDefinition.createResultKeyStatisticsCollector();
+        Set<Integer> finishedWorkers = new HashSet<>();
+
+        for (int workerNo : workerIdsWithTimeChunk) {
+          executorService.submit(() -> {
+            try {
+              ClusterByStatisticsSnapshot singletonStatisticsSnapshot =
+                  workerClient.fetchSingletonStatisticsSnapshot(
+                      workerTaskIds.get(workerNo),
+                      stageDefinition.getId().getQueryId(),
+                      stageDefinition.getStageNumber(),
+                      timeChunk
+                  );
+              // If the future already failed for some reason, stop the task.
+              if (partitionFuture.isDone()) {
+                return;
+              }
+              synchronized (mergedStatisticsCollector) {
+                mergedStatisticsCollector.addAll(singletonStatisticsSnapshot);
+                finishedWorkers.add(workerNo);
+
+                if (finishedWorkers.size() == workerIdsWithTimeChunk.size()) {
+                  Either<Long, ClusterByPartitions> longClusterByPartitionsEither =
+                      stageDefinition.generatePartitionsForShuffle(mergedStatisticsCollector);
+
+                  if (longClusterByPartitionsEither.isError()) {
+                    partitionFuture.complete(longClusterByPartitionsEither);
+                  }
+
+                  List<ClusterByPartition> timeSketchPartitions =
+                      stageDefinition.generatePartitionsForShuffle(mergedStatisticsCollector)
+                                     .valueOrThrow()
+                                     .ranges();
+                  abutAndAppendPartitionBoundries(finalPartitionBoundries, timeSketchPartitions);
+
+                  submitFetchingTasksForNextTimeChunk();
+                }
+              }
+            }
+            catch (Exception e) {
+              partitionFuture.completeExceptionally(e);
+            }
+          });
+        }
+      }
+    }
+
+    private void abutAndAppendPartitionBoundries(
```

Comment:
nit: Please add a javadoc for this method. Are there any preconditions on the arguments of this method for it to work as intended? (I am thinking something like `timeSketchPartitions` should be ordered s.t they are in an 'ascending' order. Also what if there's a gap in between the `timeSketchPartitions`'s boundaries? If that is disallowed then maybe we should add that as well).

Reply:
Added